dbgv/dbgv_test.go

```diff
@@ -0,0 +1,735 @@
+package dbgv
+
+import (
+	"fmt"
+	"runtime"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/tuneinsight/lattigo/v3/bgv"
+	"github.com/tuneinsight/lattigo/v3/drlwe"
+	"github.com/tuneinsight/lattigo/v3/ring"
+	"github.com/tuneinsight/lattigo/v3/rlwe"
+	"github.com/tuneinsight/lattigo/v3/utils"
+)
+
+var parties int = 3
+
+func testString(opname string, parties int, params bgv.Parameters) string {
+	return fmt.Sprintf("%s/LogN=%d/logQ=%d/parties=%d", opname, params.LogN(), params.LogQP(), parties)
+}
+
+var (
+	// TESTN13QP218 is a of 128-bit secure test parameters set with a 32-bit plaintext and depth 4.
+	TESTN14QP418 = bgv.ParametersLiteral{
+		LogN: 4,
+		Q:    []uint64{0x3fffffa8001, 0x1000090001, 0x10000c8001, 0x10000f0001, 0xffff00001},
+		P:    []uint64{0x7fffffd8001},
+		T:    0xffc001,
+	}
+
+	// TestParams is a set of test parameters for BGV ensuring 128 bit security in the classic setting.
+	TestParams = []bgv.ParametersLiteral{TESTN14QP418}
+)
+
+type testContext struct {
+	params bgv.Parameters
+
+	// Polynomial degree
+	n int
+
+	// Polynomial contexts
+	ringT *ring.Ring
+	ringQ *ring.Ring
+	ringP *ring.Ring
+
+	encoder bgv.Encoder
+
+	sk0Shards []*rlwe.SecretKey
+	sk0       *rlwe.SecretKey
+
+	sk1       *rlwe.SecretKey
+	sk1Shards []*rlwe.SecretKey
+
+	pk0 *rlwe.PublicKey
+	pk1 *rlwe.PublicKey
+
+	encryptorPk0 bgv.Encryptor
+	decryptorSk0 bgv.Decryptor
+	decryptorSk1 bgv.Decryptor
+	evaluator    bgv.Evaluator
+
+	crs            drlwe.CRS
+	uniformSampler *ring.UniformSampler
+}
+
+func TestDBGV(t *testing.T) {
+
+	var err error
+	for _, p := range TestParams[:] {
+
+		var params bgv.Parameters
+		if params, err = bgv.NewParametersFromLiteral(p); err != nil {
+			t.Fatal(err)
+		}
+
+		var tc *testContext
+		if tc, err = gentestContext(params); err != nil {
+			t.Fatal(err)
+		}
+		for _, testSet := range []func(tc *testContext, t *testing.T){
+
+			testPublicKeyGen,
```

Comment:
Same as for benchmark: keygen-related tests have been added to the `rlwe` package.

Reply:
done